code/mob/new_player.dm

```diff
@@ -366,9 +366,54 @@ var/global/datum/mutex/limited/latespawning = new(5 SECONDS)
 				if (istype(character.loc, /obj/machinery/vehicle))
 					boutput(character.mind.current,"<h3 class='notice'>You've become lost on your way to the station! Good luck!</h3>")
 			else if (character.traitHolder && character.traitHolder.hasTrait("sleepy"))
-				SPAWN(10 SECONDS) //ugly hardcoding- matches the duration you're asleep for
+				var/datum/trait/T = character.traitHolder.getTrait("sleepy")
+				SPAWN(T.spawn_delay)
 					boutput(character?.mind?.current,"<h3 class='notice'>Hey, you! You're finally awake!</h3>")
 				//As with the Stowaway trait, location setting is handled elsewhere.
+			else if (character.traitHolder && character.traitHolder.hasTrait("partyanimal"))
+				var/datum/trait/T = character.traitHolder.getTrait("partyanimal")
+				var/list/valid_tables = list()
+				var/list/table_turfs = list()
+
+				for_by_tcl(table, /obj/table)
+					var/is_bar = istype(get_area(table), /area/station/crew_quarters/bar) || istype(get_area(table), /area/station/crew_quarters/cafeteria)
+					if (!is_bar)
+						continue
+					if (table.z != Z_LEVEL_STATION)
+						continue
+					if ((locate(/mob/living/carbon/human) in get_turf(table)))
+						continue
+					valid_tables += table
+					table_turfs += get_turf(table)
+
+				if (length(valid_tables) > 0)
+					var/picked_table = pick(valid_tables)
+					var/starting_loc = get_turf(picked_table)
+					character.set_loc(starting_loc)
+					character.layer = 2.5 // so that they wake up under a table
+
+					var/turf/new_turf = null
+					for (var/turf/spot in range(1, character))
+
+						if (character in spot) // Skip range 0 tile
+							continue
+						if (!spot.Enter(character)) // Make sure we can walk there
+							continue
+						if ((locate(/obj/window) in spot)) // Windows can be entered, apparently :)
+							continue
```

Comment:
instead of rolling your own checks here and elsewhere, consider using `jpsTurfPassable()`